File: CLAUDE/scam-slingshot/_tools/scenarios/p1-r8-lance.mjs
/**
 * p1-r8-lance.mjs — THE r7-critic METRIC: the shape of the lance along the shot axis.
 *
 * The r7 critic measured our launch trail "tapered backwards": half-width 0.23 AD at 4–6 AD
 * from the fork growing to 0.78–0.85 AD at 12–14 AD, which is exactly where the dart is — so
 * the widest, densest end of the trail sat ON the projectile and engulfed a ~1 AD silhouette,
 * with 26–31 sparkles inside 2 AD of the ammo from t = 0 to t = 100 ms.
 *
 * Method is `p1-r6-bandvfx.mjs`'s honest isolation, extended: three renders of the SAME
 * simulated instant via SS.__render() (shipped / fx hidden / band hidden), so the VFX pixel
 * set is a DIFFERENCE and needs no colour guess, and a band buried under the burst correctly
 * counts as zero band pixels. What is new here is that it also records the PROJECTILE's screen
 * position at every timestamp, which is what lets the analyser bin the VFX along the real
 * pouch->ammo axis and say where the trail ends relative to the shot.
 *
 * Writes A/NOFX/NOBAND PNGs + geom.json. Analysed by `_tools/p1-r8-lance.py`.
 * Run it alongside `p1-r6-bandvfx.mjs` after ANY change to LANCE — the two metrics pull in
 * opposite directions (moving trail mass off the ammo moves it toward the sling, where the
 * band has to keep winning the pixel count) and neither is allowed to buy the other.
 */
export default async ({ page, game, OUT }) => {
  const say = (k, v) => console.log('### ' + k + ' ' + JSON.stringify(v));
  const TS = [0, 30, 50, 60, 100, 150];
  const ANGLE = +(process.env.P1_ANGLE ?? 0.42);
  const POWER = +(process.env.P1_POWER ?? 0.90);
  say('shot', { angle: ANGLE, power: POWER });

  await game('SS.seed(11); await SS.seek(2000);');
  await game('await SS.aim({angle:args[0], power:args[1]}); await SS.seek(400);', ANGLE, POWER);
  const info = await game('return await SS.release();');
  say('release', { ad: info.ad, muzzleAD: info.muzzleAD, power: info.power, ammo: info.ammo });

  /** Pouch + live projectile, both projected with the game's own camera, every timestamp. */
  const probe = () => game(`
    const w = SS.__world, s = w.sling, cam = w.camera;
    const r = w.renderer.domElement.getBoundingClientRect();
    const V3 = cam.position.constructor;
    const P = (x,y) => { const v = new V3(x,y,0).project(cam);
      return [ (v.x*0.5+0.5)*r.width, (-v.y*0.5+0.5)*r.height ]; };
    const halfH = Math.abs(cam.position.z) * Math.tan(cam.fov*Math.PI/360);
    const shot = w.projectiles[w.projectiles.length - 1];
    const m = shot && shot.mesh ? shot.mesh.position : null;
    return { anchorPx: P(s.anchor.x, s.anchor.y),
             ammoPx: m ? P(m.x, m.y) : null,
             ammoWorld: m ? { x: m.x, y: m.y } : null,
             pxPerWorld: (r.height/2)/halfH,
             cssW: r.width, cssH: r.height, dpr: devicePixelRatio };`);

  const setVis = (what, on) => game(`
    const w = SS.__world;
    if (args[0] === 'fx') {
      w.scene.traverse(o => { if (o.name && o.name.startsWith('fx-')) o.visible = args[1]; });
    } else {
      for (const b of w.sling.bands) b.tube.group.visible = args[1];
    }
    SS.__render();
    return true;`, what, on);

  const cap = (name) => page.screenshot({ path: `${OUT}/${name}.png` });

  const perT = {};
  let t = 0;
  for (const want of TS) {
    if (want > t) { await game('await SS.seek(args[0]);', want - t); t = want; }
    await game('SS.__render();');
    const p = await probe();
    perT[want] = { anchorPx: p.anchorPx, ammoPx: p.ammoPx, adPx: info.ad * p.pxPerWorld };
    await cap(`A-t${want}`);
    await setVis('fx', false);   await cap(`NOFX-t${want}`);   await setVis('fx', true);
    await setVis('band', false); await cap(`NOBAND-t${want}`); await setVis('band', true);
    await game('SS.__render();');
  }

  const fs = await import('node:fs/promises');
  const g0 = await probe();
  await fs.writeFile(`${OUT}/geom.json`, JSON.stringify(
    { ...g0, ts: TS, ad: info.ad, muzzleAD: info.muzzleAD, adPx: info.ad * g0.pxPerWorld,
      angle: ANGLE, power: POWER, perT }, null, 2));
  console.log('### DONE');
};

File: CLAUDE/scam-slingshot/_tools/p1-r8-lance.py
#!/usr/bin/env python3
"""
p1-r8-lance.py — scores the shape of the launch trail from the PNGs p1-r8-lance.mjs writes.

The r7 critic's gap, in numbers: our trail's half-width GREW downrange (0.23 AD at 4–6 AD from
the fork to 0.78–0.85 AD at 12–14 AD, where the dart is), so its widest and densest end sat on
the projectile, and 26–31 sparkles lived inside 2 AD of the ammo from t = 0 to t = 100 ms.
The reference is the opposite: a smear that thins downrange and STOPS short of the shot.

For each timestamp, VFX pixels are isolated by differencing the shipped frame against the same
instant with every fx-* pool hidden — no colour threshold, no guess about what is a sparkle.
They are then projected onto the live pouch->ammo axis and binned in 2-AD steps:

  n      pixels in the bin
  hw     90th-percentile |offset| from the axis, in AD  <- THE TAPER. Must DECREASE downrange.
  tip    99th-percentile reach along the axis, in AD from the pouch
  gap    ammo distance - tip, in AD                     <- must stay >= ~2 AD: the dart flies clean
  near2  VFX pixels within 2 AD of the ammo centre      <- the engulfing. Wants to be ~0.
  blobs  connected components of those                  <- "26-31 sparkles inside 2 AD"

Also prints the r5-critic band ratio (vfx:band inside a 520 px box on the pouch) at every
timestamp, because the two metrics pull in opposite directions and neither may buy the other.

Usage: python3 p1-r8-lance.py <shots-dir>
"""
import json, os, sys
import numpy as np
from PIL import Image

D = sys.argv[1]
BOX = float(sys.argv[2]) if len(sys.argv) > 2 else 520.0
g = json.load(open(os.path.join(D, 'geom.json')))


def arr(p):
    return np.asarray(Image.open(p).convert('RGB'), dtype=np.int16)


first = arr(os.path.join(D, f"A-t{g['ts'][0]}.png"))
H, W = first.shape[:2]
S = W / g['cssW']                       # device px per css px
ys, xs = np.mgrid[0:H, 0:W]

print(f"frame {W}x{H}  scale {S:g}x css   AD={g['adPx']*S:.1f} devpx   "
      f"muzzle={g['muzzleAD']} AD   shot angle={g['angle']} power={g['power']}")


def components(mask):
    """Count 8-connected blobs without scipy: iterative flood fill over the True pixels."""
    seen = np.zeros_like(mask, dtype=bool)
    idx = np.argwhere(mask)
    n = 0
    for sy, sx in idx:
        if seen[sy, sx]:
            continue
        n += 1
        stack = [(sy, sx)]
        seen[sy, sx] = True
        while stack:
            y, x = stack.pop()
            for dy in (-1, 0, 1):
                for dx in (-1, 0, 1):
                    yy, xx_ = y + dy, x + dx
                    if 0 <= yy < mask.shape[0] and 0 <= xx_ < mask.shape[1] \
                            and mask[yy, xx_] and not seen[yy, xx_]:
                        seen[yy, xx_] = True
                        stack.append((yy, xx_))
    return n


for t in g['ts']:
    e = g['perT'][str(t)] if str(t) in g['perT'] else g['perT'][t]
    px, py = e['anchorPx'][0] * S, e['anchorPx'][1] * S
    ad = e['adPx'] * S
    A = arr(os.path.join(D, f'A-t{t}.png'))
    nf = arr(os.path.join(D, f'NOFX-t{t}.png'))
    nb = arr(os.path.join(D, f'NOBAND-t{t}.png'))
    dv = (np.abs(A - nf).max(axis=2) > 8)
    db = (np.abs(A - nb).max(axis=2) > 8)

    if not e['ammoPx']:
        print(f"t={t}: no projectile"); continue
    ax_, ay_ = e['ammoPx'][0] * S, e['ammoPx'][1] * S
    vx, vy = ax_ - px, ay_ - py
    L = np.hypot(vx, vy) or 1.0
    ux, uy = vx / L, vy / L
    nx, ny = -uy, ux
    dist = L / ad                        # pouch -> ammo, in AD

    yy, xx_ = np.nonzero(dv)
    s = ((xx_ - px) * ux + (yy - py) * uy) / ad
    d = ((xx_ - px) * nx + (yy - py) * ny) / ad
    corridor = (np.abs(d) < 2.5) & (s > -0.5)
    s_c, d_c = s[corridor], d[corridor]

    # the trail's own end: ignore the pixels the ammo itself is standing in
    trail = s_c < dist - 0.6
    tip = np.percentile(s_c[trail], 99) if trail.sum() > 20 else 0.0

    near = np.hypot(xx_ - ax_, yy - ay_) < 2.0 * ad
    nearmask = np.zeros((H, W), dtype=bool)
    nearmask[yy[near], xx_[near]] = True
    blobs = components(nearmask) if near.sum() < 40000 else -1

    half = BOX / 2.0            # DEVICE px, matching p1-r6-bandvfx.py's box convention
    inbox = (np.abs(xs - px) <= half) & (np.abs(ys - py) <= half)
    bp, vp = int((db & inbox).sum()), int((dv & inbox).sum())
    ratio = (vp / bp) if bp else float('inf')

    print(f"\n--- t={t} ms   ammo at {dist:.2f} AD   tip {tip:.2f} AD   "
          f"GAP {dist - tip:5.2f} AD   near2 {int(near.sum()):5d} px / {blobs:3d} blobs   "
          f"band box vfx:band {ratio:5.2f} {'OK' if ratio < 1 else 'FAIL'}")
    print("     bin(AD)     n     hw(AD)")
    lo = 0.0
    while lo < dist + 2:
        sel = (s_c >= lo) & (s_c < lo + 2)
        if sel.sum() >= 12:
            hw = np.percentile(np.abs(d_c[sel]), 90)
            print(f"    {lo:5.1f}-{lo+2:4.1f}  {int(sel.sum()):5d}     {hw:.2f}")
        else:
            print(f"    {lo:5.1f}-{lo+2:4.1f}  {int(sel.sum()):5d}        -")
        lo += 2
